src/services/authorization/action-authorization.ts

```diff
@@ -0,0 +1,368 @@
+        collectionName,
+      });
+
+    if (!doesTriggerRequiresApproval) {
+      return false;
+    }
+
+    const requiresConditionApprovalPlainTree = await this.forestAdminClient.permissionService
+      .getConditionalRequiresApprovalCondition({
+        userId,
+        customActionName,
+        collectionName,
+      });
+
+    if (requiresConditionApprovalPlainTree) {
+      const matchingRecordsCount = await ActionAuthorizationService
+        .aggregateCountConditionIntersection(
+          recordsCounterParams,
+          filterForCaller,
+          requiresConditionApprovalPlainTree,
+        );
+
+      // No records match the condition, trigger does not require approval
+      if (matchingRecordsCount === 0) {
+        return false;
+      }
+    }
+
+    return true;
+  }
+
+  private async canApproveCustomAction(
+    userId: string | number,
+    customActionName: string,
+    collectionName: string,
+    filterForCaller: GenericPlainTree,
+    recordsCounterParams: RecordsCounterParams,
+    requesterId: number | string,
+  ): Promise<boolean> {
+    const canApprove = await this.forestAdminClient.permissionService.canApproveCustomAction({
+      userId,
+      customActionName,
+      collectionName,
+      requesterId,
+    });
+
+    if (!canApprove) {
+      return false;
+    }
+
+    const approveConditionPlainTree = await this.forestAdminClient.permissionService
+      .getConditionalApproveCondition({
+        userId,
+        customActionName,
+        collectionName,
+      });
+
+    return ActionAuthorizationService.canPerformConditionalCustomAction(
+      recordsCounterParams,
+      filterForCaller,
+      approveConditionPlainTree,
+    );
+  }
+
+  private async getRoleIdsAllowedToApprove(
+    recordsCounterParams: RecordsCounterParams,
+    customActionName: string,
+    collectionName: string,
+    filterForCaller: GenericPlainTree,
+  ) {
+    const actionConditionsByRoleId = await this.forestAdminClient.permissionService
+      .getConditionalApproveConditions({
+        customActionName,
+        collectionName,
+      });
+
+    const roleIdsAllowedToApproveWithoutConditions = await this.forestAdminClient.permissionService
+      .getRoleIdsAllowedToApproveWithoutConditions({
+        customActionName,
+        collectionName,
+      });
+
+    // Optimization - We groupBy conditions to only make the aggregate count once when possible
+    const rolesIdsGroupByConditions = ActionAuthorizationService
+      .transformToRolesIdsGroupByConditions(
+        actionConditionsByRoleId,
+      );
+
+    if (!rolesIdsGroupByConditions.length) {
+      return roleIdsAllowedToApproveWithoutConditions;
+    }
+
+    const [requestRecordsCount, ...conditionRecordsCounts] = await Promise.all([
+      ActionAuthorizationService.aggregateCountConditionIntersection({
+        ...recordsCounterParams,
+        excludesScope: true,
+      }, filterForCaller),
+      // eslint-disable-next-line max-len
```

Comment:
To remove

Reply:
Nope, the formatter keeps formatting the line as one long line :'(